Index: JmDNSImpl.java
===================================================================
RCS file: /cvsroot/jmdns/jmdns/src/javax/jmdns/impl/JmDNSImpl.java,v
retrieving revision 1.4
diff -u -r1.4 JmDNSImpl.java
--- JmDNSImpl.java	21 May 2008 21:00:23 -0000	1.4
+++ JmDNSImpl.java	4 Aug 2008 09:16:07 -0000
@@ -101,10 +101,6 @@
      * Values hold the fully qualified service type.
      */
     Map serviceTypes;
-    /**
-     * This is the shutdown hook, we registered with the java runtime.
-     */
-    private Thread shutdown;
 
     /**
      * Handle on the local host
@@ -253,8 +249,6 @@
         // I would pass' JmDNS.Timer'.
         timer = new Timer();
         new RecordReaper(this).start(timer);
-        shutdown = new Thread(new Shutdown(), "JmDNS.Shutdown");
-        Runtime.getRuntime().addShutdownHook(shutdown);
 
         incomingListener = new Thread(new SocketListener(this), "JmDNS.SocketListener");
 
@@ -1146,19 +1140,6 @@
         timer.schedule(task, delay);
     }
 
-    // REMIND: Why is this not an anonymous inner class?
-    /**
-     * Shutdown operations.
-     */
-    private class Shutdown implements Runnable
-    {
-        public void run()
-        {
-            shutdown = null;
-            close();
-        }
-    }
-
     /**
      * Recover jmdns when there is an error.
      */
@@ -1230,13 +1211,6 @@
 
                 // Stop the timer
                 timer.cancel();
-
-                // remove the shutdown hook
-                if (shutdown != null)
-                {
-                    Runtime.getRuntime().removeShutdownHook(shutdown);
-                }
-
             }
         }
     }

